liquid-compiler/src/parser.rs

```diff
@@ -3,477 +3,831 @@
+        panic!("Expected a filter.");
+    }
 
-        match iter.peek() {
-            Some(&&Token::Pipe) | None => {
-                filters.push(FilterCall::new(name, args));
-                continue;
-            }
-            _ => (),
+    let mut filter = filter.into_inner();
+    let name = filter.next().expect("A filter always has a name.").as_str();
+    let args = filter.map(parse_value).collect();
+
+    FilterCall::new(name, args)
+}
+
+/// Parses a `FilterChain` from a `Pair` with a filter chain.
+/// This `Pair` must be `Rule::FilterChain`.
+fn parse_filter_chain(chain: Pair) -> FilterChain {
+    if chain.as_rule() != Rule::FilterChain {
+        panic!("Expected an expression with filters.");
+    }
+
+    let mut chain = chain.into_inner();
+    let entry = parse_value(
+        chain
+            .next()
+            .expect("A filterchain always has starts by a value."),
+    );
+    let filters = chain.map(parse_filter).collect();
+
+    FilterChain::new(entry, filters)
+}
+
+/// An interface to access elements inside a block.
+pub struct TagBlock<'a: 'b, 'b> {
+    name: &'b str,
+    iter: &'b mut Iterator<Item = Pair<'a>>,
+    closed: bool,
+}
+
+impl<'a, 'b> TagBlock<'a, 'b> {
+    fn new(name: &'b str, next_elements: &'b mut Iterator<Item = Pair<'a>>) -> Self {
+        TagBlock {
+            name,
+            iter: next_elements,
+            closed: false,
+        }
+    }
+
+    /// Returns the next element of the block, if any, similarly to an iterator.
+    ///
+    /// However, if the input text reaches its end and the block is not closed,
+    /// an error is returned instead.
+    pub fn next(&mut self) -> Result<Option<BlockElement<'a>>> {
+        if self.closed {
+            return Ok(None);
+        }
+
+        let element = self.iter.next().expect("File shouldn't end before EOI.");
+
+        if element.as_rule() == Rule::EOI {
+            return Err(error_from_pair(
+                element,
+                format!("Unclosed block. {{% end{} %}} tag expected.", self.name),
+            ));
         }
 
-        expect(&mut iter, &Token::Colon)?;
-
-        // loops through the argument list after the filter name
-        while iter.peek() != None && iter.peek().unwrap() != &&Token::Pipe {
-            args.push(iter.next().unwrap().to_arg()?);
-
-            // ensure that the next token is either a Comma or a Pipe
-            match iter.peek() {
-                Some(&&Token::Comma) => {
-                    let _ = iter.next().unwrap();
-                    continue;
-                }
-                Some(&&Token::Pipe) | None => break,
-                _ => {
-                    return Err(unexpected_token_error(
-                        "`,` | `|`",
-                        Some(iter.next().unwrap()),
-                    ));
-                }
+        if element.as_rule() == Rule::Tag {
+            let tag_name = element
+                .clone()
```

Comment:
Do we need this clone?

Reply:
We did in the past. Now that I look at it, with `Tag` being a structure that already separates the name from tokens, it might be possible to get rid of it, though.